template-manager-cmtemplate/src/main/java/com/sequenceiq/cloudbreak/cmtemplate/configproviders/impala/ImpalaVolumeConfigProvider.java

```diff
@@ -38,20 +40,51 @@ public List<ApiClusterTemplateConfig> getRoleConfigs(String roleType, HostgroupV
                         .mapToInt(volume -> volume.getVolumeSize())
                         .min()
                         .orElse(0);
+                Integer volumeCount = 0;
+                Integer temporaryStorageVolumeCount = 0;
+                Integer temporaryStorageVolumeSize = 0;
+                if (hostGroupView != null) {
+                    volumeCount = hostGroupView.getVolumeCount();
+                    if (hostGroupView.getTemporaryStorageVolumeCount() != null) {
+                        temporaryStorageVolumeCount = hostGroupView.getTemporaryStorageVolumeCount();
+                    }
+                    if (hostGroupView.getTemporaryStorageVolumeSize() != null) {
+                        temporaryStorageVolumeSize = hostGroupView.getTemporaryStorageVolumeSize();
+                    }
+                }
 
                 List<ApiClusterTemplateConfig> configs = new ArrayList<>();
+
+                if (checkTemporaryStorage(hostGroupView, temporaryStorageVolumeCount)) {
+                    configs.add(config(IMPALA_SCRATCH_DIRS_PARAM,
+                            buildEphemeralVolumePathString(temporaryStorageVolumeCount, "impala/scratch")));
+                } else {
                 configs.add(config(IMPALA_SCRATCH_DIRS_PARAM,
-                        buildVolumePathStringZeroVolumeHandled(hostGroupView.getVolumeCount(), "impala/scratch")));
+                        buildVolumePathStringZeroVolumeHandled(volumeCount, "impala/scratch")));
+                }
 
                 if (minAttachedVolumeSize > MIN_ATTACHED_VOLUME_SIZE_TO_ENABLE_CACHE_IN_GB
                         && hostGroupView.getVolumeCount() > 0) {
                     long dataCacheCapacityInBytes = VolumeUtils
                             .convertGBToBytes(Math.min(minAttachedVolumeSize / 2, MAX_IMPALA_DATA_CACHE_SIZE_IN_GB / hostGroupView.getVolumeCount()));
 
                     configs.add(config(IMPALA_DATACACHE_ENABLED_PARAM, "true"));
-                    configs.add(config(IMPALA_DATACACHE_CAPACITY_PARAM, Long.toString(dataCacheCapacityInBytes)));
-                    configs.add(config(IMPALA_DATACACHE_DIRS_PARAM,
-                            buildVolumePathStringZeroVolumeHandled(hostGroupView.getVolumeCount(), "impala/datacache")));
+
+                    if (checkTemporaryStorage(hostGroupView, temporaryStorageVolumeCount)) {
+                        long temporaryStoragedataCacheCapacityInBytes = VolumeUtils
+                                .convertGBToBytes(Math.min(temporaryStorageVolumeSize / 2,
+                                        MAX_IMPALA_DATA_CACHE_SIZE_IN_GB / hostGroupView.getTemporaryStorageVolumeCount()));
```

Comment:
why not using the local variable?

Reply:
Yes, now local variable can be used. First, I had not created the local variable. So, I was calling it directly. I will change it.